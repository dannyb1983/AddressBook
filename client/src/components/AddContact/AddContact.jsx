//modules
import React, { Fragment, useState } from "react";
import { Form, Col, Row, Container, Button } from "react-bootstrap";
import { useHistory } from "react-router-dom";
import { v4 as uuidv4 } from "uuid";

const AddContact = ({ addContact, sameFirstLastNameCheck }) => {
  const [canAddContact, setCanAddContact] = useState(null);
  const history = useHistory();
  const contactPanel = ["First Name", "Last Name", "Email", "Phone"];

  // add a new Contact
  const onFormSubmit = (e) => {
    e.preventDefault();
    const newContact = {};
    newContact.login = { uuid: uuidv4() };
    newContact.name = { first: e.target[0].value, last: e.target[1].value };
    newContact.email = e.target[2].value;
    newContact.phone = e.target[3].value;
    newContact.picture = { thumbNail: "../../../../assets/noimage.jpg" };
    // if first or last name is left blank
    if (newContact.name.first.length < 1 || newContact.name.last.length < 1) {
      console.log("New contacts must have a first AND last name!");
      setCanAddContact(false);
      return;
    } // check that the first and last name are unique
    if (sameFirstLastNameCheck(newContact)) {
      let sameUserConfirmation = window.confirm(
        "A contact with this name already exists. Are you sure you want to continue?"
      );
      if (!sameUserConfirmation) {
        return;
      } else {
        addContact(newContact);
        setCanAddContact(true);
        history.push("/");
      }
    } else {
      addContact(newContact);
      setCanAddContact(true);
      history.push("/");
    }
  };
  const goBack = () => {
    history.push("/");
  };

  return (
    <div title="Click to show details">
      <h1>Add a new Contact</h1>
      <Container className="p-3">
        <Row>
          <Col md={12} className="text-left py-5 mt-3 ml-0 pl-0">
            <div style={{ paddingLeft: "30px" }}>
              <Button variant="secondary" onClick={goBack}>
                <i className="fas fa-arrow-left mr-2"></i>back
              </Button>
            </div>
          </Col>
        </Row>
        <Fragment>
          <Form onSubmit={onFormSubmit}>
            {contactPanel.map((field) => {
              return (
                <Form.Group
                  key={`AddContactField_${field}`}
                  as={Row}
                  className="mb-3"
                  controlId="formPlaintextEmail"
                >
                  <Form.Label column sm="2">
                    {field}
                  </Form.Label>
                  <Col sm="10">
                    <Form.Control placeholder={field} type="text" />
                  </Col>
                </Form.Group>
              );
            })}
            <div style={{ padding: "10px", textAlign: "center" }}>
              <Button variant="primary" type="submit">
                Submit
              </Button>
            </div>
          </Form>
          {canAddContact === false ? (
            <h5 style={{ color: "red", padding: "20px", textAlign: "center" }}>
              **Contacts must contain a first AND last name
            </h5>
          ) : null}
        </Fragment>
      </Container>
    </div>
  );
};

export default AddContact;
