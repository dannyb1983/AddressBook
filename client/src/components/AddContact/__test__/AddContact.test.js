//modules
import React from 'react';
import {render, screen,cleanup} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { contact, contacts } from '../../../../../test/jest/__mocks__/testData';
//components
import AddContact from '../AddContact';

describe('Add Contact (Unit)', () => {
  afterEach(() => cleanup());

  const renderAddContact = () => render(<MemoryRouter><AddContact addContact={() => contacts} sameFirstLastNameCheck={() => false} onFormSubmit={() => contact} /></MemoryRouter>);
  const renderEmptyAddContact = () =>render(<MemoryRouter><AddContact canAddContact='false' /></MemoryRouter>);

  describe('render()', () => {
    const { asFragment, } = renderAddContact();
    
    it('AddContact should match snapshot', () => {
      expect(asFragment().firstChild).toMatchSnapshot();
    });  

    it('Should render all form fields', () => {
      renderAddContact();
      expect(screen.getByText('First Name')).toBeInTheDocument();
      expect(screen.getByText('Last Name')).toBeInTheDocument();
      expect(screen.getByText('Email')).toBeInTheDocument();
      expect(screen.getByText('Phone')).toBeInTheDocument();
    });
  });
});
