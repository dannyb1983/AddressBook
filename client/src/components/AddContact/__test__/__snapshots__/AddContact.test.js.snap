// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Contact (Unit) render() AddContact should match snapshot 1`] = `
<div
  title="Click to show details"
>
  <h1>
    Add a new Contact
  </h1>
  <div
    class="p-3 container"
  >
    <div
      class="row"
    >
      <div
        class="text-left py-5 mt-3 ml-0 pl-0 col-md-12"
      >
        <div
          style="padding-left: 30px;"
        >
          <button
            class="btn btn-secondary"
            type="button"
          >
            <i
              class="fas fa-arrow-left mr-2"
            />
            back
          </button>
        </div>
      </div>
    </div>
    <form
      class=""
    >
      <div
        class="mb-3 form-group row"
      >
        <label
          class="form-label col-form-label col-sm-2"
          for="formPlaintextEmail"
        >
          First Name
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="formPlaintextEmail"
            placeholder="First Name"
            type="text"
          />
        </div>
      </div>
      <div
        class="mb-3 form-group row"
      >
        <label
          class="form-label col-form-label col-sm-2"
          for="formPlaintextEmail"
        >
          Last Name
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="formPlaintextEmail"
            placeholder="Last Name"
            type="text"
          />
        </div>
      </div>
      <div
        class="mb-3 form-group row"
      >
        <label
          class="form-label col-form-label col-sm-2"
          for="formPlaintextEmail"
        >
          Email
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="formPlaintextEmail"
            placeholder="Email"
            type="text"
          />
        </div>
      </div>
      <div
        class="mb-3 form-group row"
      >
        <label
          class="form-label col-form-label col-sm-2"
          for="formPlaintextEmail"
        >
          Phone
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="formPlaintextEmail"
            placeholder="Phone"
            type="text"
          />
        </div>
      </div>
      <div
        style="padding: 10px; text-align: center;"
      >
        <button
          class="btn btn-primary"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;
